### DISCUSSION TITLE
easily access members of inherited class from instance of base class without explicit casting

### DISCUSSION THREAD

**Discussion by Caleb-D** [dotnet/csharplang#7048] [Language Ideas]
```
IData
{
  // some genericly named members to access data (think vectors + metadata)
  object GetVal(string name);
}

IFriendly1 : IData
{
  // some members in the terminology of use case 1 where data is of format 1
  int FriendlyInfo1 { get { return GetVal("info1"); } }
  double FriendlyInfo2; { get { return GetVal("info2"); } }
}

IFriendly2 : IData
{
  // another interface, not in this assembly and only known at runtime (think plugin)
  DateTime FriendlyInfo3 {get {...}}
}

IActivity
{
  IReadOnlyList<IData> TheData;
}

var activity = get.the.activity();
var result = activity.TheData[0];
// now user has to go read documentation...oh, there's a GetVal, what string do I use??  later, oh, there's IFriendly1, I will cast.
var friendly1 = result as IFriendly1;
// user happily uses friendly1, but it took a while to discover how to do this
var a = friendly1.FriendlyInfo1;
```
Now, imagine if all IFriendly were included in assembly references (not usings yet) and the user typed:
`var a = result.`
and the intellisense discovered (because it is part of the language, not because it's a super intellisense) that there are two interfaces that could be used.  Then you see autocomplete for all the members of all the interfaces.

Then you could complete 
`var a = result.FriendlyInfo1; `
and the compiler would automatically cast behind the scenes.  A using statement would be added by the intellisense completion if needed.  

If there are multiple interfaces with a member that is the same name, the using statement would limit it, or the autocomplete could be something like
`var a = result.(SameName from IFriendly3);`

At runtime, if the type were of some other interface, you would get a type cast exception.  Or you could use ? to get a nullable:
`var a = result.(SameName? from IFriendly3);`

**Comment by RikkiGibson** [comment]
I'm not sure the language/tooling will want to enshrine this practice--discovering subtypes, suggesting their members in lookup, and inserting casts automatically.

If this is really the way your type will end up getting used, you might want to add an extension to `IData` like `AsIFriendly1()` which converts it to either a Friendly1 or null, so the user can discover that extension in lookup, and go from there.

**Comment by Caleb-D** [comment]
@RikkiGibson thanks for the suggestion.  Yes, you're probably right...
